### ISSUE TITLE
[ADAP-1044] [Bug] Incremental model with merge strategy fails to remove __dbt_tmp table when using on_schema_change=sync_all_columns 

### ISSUE THREAD

**Issue by MikkoPaajanen** [dbt-labs/dbt-bigquery#1032]
### Is this a new bug in dbt-bigquery?

- [X] I believe this is a new bug in dbt-bigquery
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Current behaviour:
adapter does not remove __dbt_tmp table when using `+on_schema_change: "sync_all_columns"` and incremental strategy "merge".

I know these are created as "temp" tables but these are not actually bigquery temp tables as stated in: https://github.com/dbt-labs/dbt-bigquery/issues/163#issuecomment-1110815079 

### Expected Behavior

After running merge script it should check if following are true (tmp_relation_exists and strategy = "merge") and then drop tmp relation.

I believe this:
```
{%- if strategy == 'merge' and tmp_relation -%}
      {{ adapter.drop_relation(tmp_relation) }}
{%- endif -%}
```
Would resolve the issue if added in : https://github.com/dbt-labs/dbt-bigquery/blob/0c5422c6df8f542611ee04e0c59c64ca0e0335a1/dbt/include/bigquery/macros/materializations/incremental.sql#L157

I don't see any reason why temp table shouldn't be dropped after the merge. Is there a reason?

### Steps To Reproduce

1. dbt-core== 1.7.1, dbt-bigquery==1.7.1
2. Configurations
In the model:
`{{ config(materialized='incremental') }}`
In dbt_project.yml:
`+on_schema_change: "sync_all_columns"`
3. Run dbt run -s <model>
4. See __dbt_tmp table in bigquery after run

### Environment

```markdown
- OS: macOS Sonoma 14.0
- Python: 3.10
- dbt-core: 1.7.1
- dbt-bigquery: 1.7.1
```

### Additional Context

I did find some discussion on the topic in other closed issues:
https://github.com/dbt-labs/dbt-bigquery/issues/163
https://github.com/dbt-labs/dbt-bigquery/issues/154

**Comment by dbeatty10** [comment]
> I don't see any reason why temp table shouldn't be dropped after the merge. Is there a reason?

Thanks for reporting this @MikkoPaajanen !

I don't see any reason why that table shouldn't be dropped after the merge. In fact, it seems to me that _any_ intermediate tables should be dropped as a cleanup step by the end of the materialization.

### Acceptance criteria

- No matter if a SQL model or a Python model, any incremental materialization that creates an intermediate table (whether it is [temporary or permanent](https://cloud.google.com/bigquery/docs/writing-results#temporary_and_permanent_tables)) table should explicitly clean up that table by the end of the materialization.

### Bonus criteria

Each of these is optional and ordered from the most impactful to least.

1. Create a functional test that verifies the intermediate table does not exist
    - e.g., use something like [`make_temp_relation`](https://github.com/dbt-labs/dbt-core/blob/c2bc2f009bbeeb46b3c69d082ab4d485597898af/core/dbt/include/global_project/macros/adapters/relation.sql#L13-L19) to determine what the intermediate object [will be named](https://github.com/dbt-labs/dbt-bigquery/blob/06851679f75d18ece98c95d4eb2a0ddd16544f4d/dbt/include/bigquery/macros/materializations/incremental.sql#L71), and make sure it doesn't exist after building the model 
3. Refactor to use [`make_intermediate_relation`](https://github.com/dbt-labs/dbt-core/blob/c2bc2f009bbeeb46b3c69d082ab4d485597898af/core/dbt/include/global_project/macros/adapters/relation.sql#L5-L7) instead of `make_temp_relation` since they are functionally equivalent, and calling it an "intermediate" relation avoids confusion with [non-permanent temporary tables in BigQuery](https://cloud.google.com/bigquery/docs/writing-results#temporary_and_permanent_tables)
4. Consider renaming `tmp_relation` to something like `intermediate_relation` for the same reasons as above

**Comment by dbeatty10** [comment]
### Implementation idea

To do this most robustly, I'd think we'd want to refactor these lines:
https://github.com/dbt-labs/dbt-bigquery/blob/06851679f75d18ece98c95d4eb2a0ddd16544f4d/dbt/include/bigquery/macros/materializations/incremental.sql#L154-L156

i.e, remove the lines above and add the code below [somewhere in here](https://github.com/dbt-labs/dbt-bigquery/blob/06851679f75d18ece98c95d4eb2a0ddd16544f4d/dbt/include/bigquery/macros/materializations/incremental.sql#L160-L167):

```sql
  {%- if tmp_relation_exists -%} 
    {{ adapter.drop_relation(tmp_relation) }} 
  {%- endif -%} 
```

**Comment by nathaniel-may** [comment]
Closing in favor of https://github.com/dbt-labs/dbt-bigquery/issues/1036.

Please re-open this issue if you can observe the temp tables persisting for more than 12 hours.

**Comment by dbeatty10** [comment]
Re-labeling as closed in favor of #1036 (gray "closed") rather than "closed as completed / resolved" (purple "closed").
